File: services/portfolio-analytics/app/ai/identity.py
import re
from typing import Optional


def score_identity(
  document_type: str,
  name_on_document: str,
  date_of_birth: Optional[str] = None,
  id_number: Optional[str] = None,
) -> dict:
  score = 0.0
  checks = []
  if document_type and document_type.lower() in ("passport", "id_card", "drivers_license"):
    score += 0.25
    checks.append("document_type_recognized")
  if name_on_document and len(name_on_document.strip()) >= 4 and re.match(r"^[\w\s\-\.]+$", name_on_document.strip()):
    score += 0.25
    checks.append("name_format_valid")
  if date_of_birth and re.match(r"^\d{4}-\d{2}-\d{2}$", date_of_birth.strip()):
    score += 0.25
    checks.append("dob_format_valid")
  if id_number and len(id_number.strip()) >= 6:
    score += 0.25
    checks.append("id_number_present")
  return {
    "identity_score": min(score, 1.0),
    "checks": checks,
    "document_type": document_type,
  }

File: apps/server-go/internal/domain/identity.go
package domain

import (
	"math"
	"regexp"
	"strings"
)

var (
	nameRegex   = regexp.MustCompile(`^[\w\s\-\.]+$`)
	dobRegex    = regexp.MustCompile(`^\d{4}-\d{2}-\d{2}$`)
	docTypes    = map[string]bool{"passport": true, "id_card": true, "drivers_license": true}
)

func IdentityKYCTier(score float64) string {
	if score >= 0.75 {
		return "high"
	}
	if score >= 0.5 {
		return "medium"
	}
	return "low"
}

func FraudRecommendation(isAnomaly bool, anomalyScore float64) string {
	if isAnomaly && anomalyScore < -0.5 {
		return "block"
	}
	if isAnomaly {
		return "review"
	}
	return "allow"
}

func SurveillanceAlertType(isAnomaly bool, quantityZscore, notionalZscore float64) string {
	const threshold = 2.5
	if !isAnomaly {
		return "none"
	}
	qBreach := math.Abs(quantityZscore) > threshold
	nBreach := math.Abs(notionalZscore) > threshold
	if qBreach && nBreach {
		return "volume_and_notional_deviation"
	}
	if qBreach {
		return "volume_spike"
	}
	return "notional_deviation"
}

type IdentityScoreResult struct {
	IdentityScore float64  `json:"identity_score"`
	KYCTier       string   `json:"kyc_tier"`
	Checks        []string `json:"checks"`
	DocumentType  string   `json:"document_type"`
}

func ScoreIdentity(documentType, nameOnDocument, dateOfBirth, idNumber string) IdentityScoreResult {
	var score float64
	var checks []string
	if documentType != "" && docTypes[strings.ToLower(documentType)] {
		score += 0.25
		checks = append(checks, "document_type_recognized")
	}
	name := strings.TrimSpace(nameOnDocument)
	if name != "" && len(name) >= 4 && nameRegex.MatchString(name) {
		score += 0.25
		checks = append(checks, "name_format_valid")
	}
	dob := strings.TrimSpace(dateOfBirth)
	if dob != "" && dobRegex.MatchString(dob) {
		score += 0.25
		checks = append(checks, "dob_format_valid")
	}
	id := strings.TrimSpace(idNumber)
	if id != "" && len(id) >= 6 {
		score += 0.25
		checks = append(checks, "id_number_present")
	}
	if score > 1.0 {
		score = 1.0
	}
	return IdentityScoreResult{
		IdentityScore: score,
		KYCTier:       IdentityKYCTier(score),
		Checks:        checks,
		DocumentType:  documentType,
	}
}
